internal/runner/minio_test.go

```diff
@@ -132,13 +198,219 @@ func s3AnonAccess(t *testing.T, mts *minio_local.MinioTestServer, logger *log.Lo
+			t.Fatal("unexpected key was downloaded", file.Name(), "dir", tmpDir, "requested", test.size, "keys", spew.Sdump(keys))
+		}
+		delete(keys, file.Name())
+
+		if errGo = os.Remove(filepath.Join(tmpDir, file.Name())); errGo != nil {
+			t.Fatal(errGo)
+		}
+	}
+	if len(keys) != 0 {
+		t.Fatal("key(s) not downloaded", keys)
+	}
+
+	if files, errGo = os.ReadDir(tmpDir); errGo != nil {
+		t.Fatal(errGo)
+	}
+	for _, file := range files {
+		os.Remove(filepath.Join(tmpDir, file.Name()))
+	}
+
+	// Now try the gather to exercise our negative test of trying to download too much, the prefix
+	// is not needed as we are using the entire bucket
+	size, warnings, _ := authS3.Gather(ctx, "", tmpDir, test.size, nil, true)
+	// err is ignored as we will go over budgets in testing to get our negative cases
+	if size > test.size {
+		t.Fatal("it appears Gather downloaded too much data", "actual", size, "expected", test.size)
+	}
+
+	if len(warnings) != 0 {
+		t.Fatal(warnings)
+	}
+
+	// Release the keys from the test specification as the first pass
+	// erases them intentionally
+	for _, key := range test.keys {
+		keys[key] = 0
+	}
+
+	if files, errGo = os.ReadDir(tmpDir); errGo != nil {
+		t.Fatal(errGo)
+	}
+
+	for _, file := range files {
+		if _, isPresent := keys[file.Name()]; !isPresent {
+			t.Fatal("unexpected key was downloaded", file.Name(), "dir", tmpDir, "requested", test.size, "keys", spew.Sdump(keys))
+		}
+		delete(keys, file.Name())
+
+		if errGo = os.Remove(filepath.Join(tmpDir, file.Name())); errGo != nil {
+			t.Fatal(errGo)
+		}
+	}
+	if len(keys) != 0 {
+		t.Fatal("key(s) not downloaded", keys)
+	}
+}
+
+func s3Limiter(t *testing.T, mts *minio_local.MinioTestServer, logger *log.Logger) {
+	// Check that the minio local server has initialized before continuing
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s3AliveCheck(ctx, t, mts)
+	if t.Failed() {
+		return
+	}
+
+	// Upload several files to the S3 servera
+	bucketsAndBlobs, err := s3TestFiles(ctx, mts, 2, 8)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Use the Storage API to connect to the server
+	creds := request.AWSCredential{
+		AccessKey: mts.AccessKeyId,
+		SecretKey: mts.SecretAccessKeyId,
+	}
+
+	tests := []testCase{}
+
+	// Use the fetch calls to test limitations
+	for _, bucket := range bucketsAndBlobs {
+		// Generate a test case where we choose and blob to stop in at random
+		// and a test case where we do not get to load any blobs, and one where
+		// we load all blobs
+		blobsTotal := int64(0)
+		blobSizes := make([]int64, 0, len(bucket.blobs))
+		keys := make([]string, 0, len(bucket.blobs))
+		for _, blob := range bucket.blobs {
+			blobSizes = append(blobSizes, blob.size)
+			blobsTotal += blob.size
+			keys = append(keys, blob.key)
+		}
+		// No blobs should be downloaded when the space is too small
+		tests = append(tests, testCase{
+			bucket: bucket.name,
+			size:   blobSizes[0] / 2,
+			keys:   []string{},
```

Comment:
So we guarantee that blob[0] has the smallest size?

Reply:
Yes I let it panic if this is wrong, being test code and all.  In general the make would have the benefit of the hard coded parameters used in the s3TestFiles.